File: bin/model_funcs/dflux.py
# python implementation of dflux.f

from operator import pos
from bin.Field import Field, is_field, maximum, minimum, abs, pos_diff, isscalar, max
from bin.Grid import Grid
from bin.Workspace import Workspace
import numpy as np

def dflux(model, ws, state, dw, rfil):
    """
    calculate artificial dissipation fluxes on finest mesh using blended first and 
    third order fluxes
    
    Args:
        model (NavierStokes): physics model
        workspace (Workspace): contains the relevant Fields
        state (Field): density, x-momentum, y-momentum, and energy
        dw (Field): to store new residuals after completing fluxes 
        rfil (float): relaxation factor determining balance between viscous and artificial dissipation fluxes
        
    """

    # take a workspace ws and calculate dissipative fluxes

    # model parameters
    pad = model.padding
    n = Field.dim(state) # number of quantities being convected

    # necessary fields
    def mget(varName):
        return ws.get_field(varName, model.className)
    w = state # state vector
    fw   = mget('fw')
    p    = mget('p')   # pressure
    porI = mget('porI')
    porJ = mget('porJ') # porosity
    radI = mget('radI')
    radJ = mget('radJ')

    # grid dimensions
    [nx, ny] = ws.field_size()
    [nxp, nyp] = [pad+nx+pad, pad+ny+pad]
    ip = pad
    jp = pad
    il = nx+1
    jl = ny+1
    ie = nx+2
    je = ny+2
    ib = nx+3
    jb = ny+3

    # physical paramteters
    vis2 = model.params['vis2']
    vis4 = model.params['vis4']

    fis2      = 2*rfil*vis2
    fis4      = rfil*vis4/16
    sfil      = 1  -rfil
    plim      = 0.001
    tol       = 0.25

    # working arrays
    dp = Field((nxp,nyp))
    dis2 = Field((nxp, nyp))
    dis4 = Field((nxp, nyp))
    d    = Field((nxp, nyp, n))
    e    = Field((nxp, nyp, n))
    TOL  = Field((nx+1, ny+1), tol)

    # c
    # c     replace the energy by the enthalpy
    # c
    w[:,:,3] = w[:,:,3] + p

    # c
    # c     dissipation in the i direction
    # c
    dp[1:ib, jp:je] = abs((p[ip:nxp, jp:je] - 2*p[1:ib, jp:je] + p[0:ie, jp:je]) / \
                          (p[ip:nxp, jp:je] + 2*p[1:ib, jp:je] + p[0:ie, jp:je] + plim))

    dp[0, jp:je] = dp[1, jp:je]
    dp[ib, jp:je] = dp[ie, jp:je]

    rad = minimum(radI[ip:ib, jp:je], radI[1:ie, jp:je])

    max1 = maximum(dp[ip+1:nxp, jp:je], dp[ip:ib, jp:je])
    max2 = maximum(dp[1:ie, jp:je], dp[0:il, jp:je])
    dpsafe = minimum(TOL[:,0:ny], maximum(max1, max2))
    dis2[1:ie, jp:je] = dpsafe*rad*fis2

    dis4[1:ie, jp:je] = rad*fis4
    dis4 = pos_diff(dis4, dis2)

    # forward differencing (1st order)
    d[0:ib, jp:je] = w[1:nxp, jp:je] - w[0:ib, jp:je]

    # central differencing (third order)
    e[1:ie, jp:je] = d[ip:ib, jp:je] - 2*d[1:ie, jp:je] + d[0:il, jp:je]

    gs = porI*(dis2[1:ie, jp:je]*d[1:ie, jp:je] - dis4[1:ie, jp:je]*e[1:ie, jp:je])

    fw[ip:ie, jp:je] = fw[ip:ie, jp:je]*sfil + gs[0:nx] - gs[1:il]

    # c
    # c     dissipation in the j direction
    # c
    if ny >= 3:
        dp[ip:ie, 1:jb] = abs((p[ip:ie, jp:nyp] - 2*p[ip:ie, 1:jb] + p[ip:ie, 0:je]) / \
                              (p[ip:ie, jp:nyp] + 2*p[ip:ie, 1:jb] + p[ip:ie, 0:je] + plim))

        
        rad = minimum(radJ[ip:ie, 2], radJ[ip:ie, 1])

        max1 = maximum(dp[ip:ie, 3], dp[ip:ie, 2])
        max2 = dp[ip:ie, 1]
        dis2[ip:ie, 1] = rad*fis2*minimum(TOL[0:nx,0], maximum(max1, max2))

        dis4[ip:ie, 1] = rad*fis4
        dis4[ip:ie, 1] = pos_diff(dis4[ip:ie, 1], dis2[ip:ie, 1])


        rad = minimum(radJ[ip:ie, je], radJ[ip:ie, jl])

        max1 = maximum(dp[ip:ie, je], dp[ip:ie, jl])
        max2 = dp[ip:ie, ny]
        dis2[ip:ie, jl] = rad*fis2*minimum(TOL[0:nx,0], maximum(max1, max2))

        dis4[ip:ie, jl] = rad*fis4
        dis4[ip:ie, jl] = pos_diff(dis4[ip:ie, jl], dis2[ip:ie, jl])

        rad = min(radJ[ip:ie, jp+1:je], radJ[ip:ie, jp:jl])

        max1 = maximum(dp[ip:ie, jp+2:jb], dp[ip:ie, jp+1:je])
        max2 = maximum(dp[ip:ie, jp:jl], dp[ip:ie, 1:ny])
        dis2[ip:ie, jp:jl] = rad*fis2*minimum(TOL[0:nx,0:ny-1], maximum(max1, max2))

        dis4[ip:ie, jp:jl] = rad*fis4
        dis4[ip:ie, jp:jl] = pos_diff(dis4[ip:ie, jp:jl], dis2[ip:ie, jp:jl])


        # forward difference (first order)
        d[ip:ie, 0:jb] = w[ip:ie, 1:nyp] - w[ip:ie, 0:jb]

        # central difference (third order)
        e[ip:ie, 1:je] = d[ip:ie, jp:jb] - 2*d[ip:ie, 1:je] + d[ip:ie, 0:jl]

        gs = porJ*(dis2[ip:ie, 1:je]*d[ip:ie, 1:je] - dis4[ip:ie, 1:je]*e[ip:ie, 1:je])

        fw[ip:ie, jp:je] += gs[:, 0:ny] - gs[:, 1:jl]

    # c
    # c     replace the enthalpy by the energy
    # c
    w[:,:,3] = w[:,:,3] - p

    # add to flux field
    dw += fw

File: bin/model_funcs/fortran_versions/dflux.f
      subroutine dflux(ny,il,jl,ie,je,ib,jb,
     & w,p,
     & pori, porj,
     & fw, radi, radj,
c    no flo_params used
     & rfil,vis2,vis4)
c     
c     arg list above is:
c
c     dims
c
c     flow_var
c     mesh_var
c     solv_var
c
c     flo_param
c     solv_param
c
c
c
c     ******************************************************************
c     *                                                                *
c     *   upwind biasing by artificial dissipation using               *
c     *   blended first and third order fluxes                         *
c     *                                                                *
c     ******************************************************************
c
c     w(i,j,1)  = density
c     w(i,j,2)  = momentum in x direction
c     w(i,j,3)  = momentum in y direction
c     w(i,j,4)  = total energy
c
c     ******************************************************************
c
c      use dims
c
c     ******************************************************************
c
c      use flo_var
c      use mesh_var
c      use solv_var
c
c     ******************************************************************
c
c      use flo_param
c      use solv_param
c
c     ******************************************************************
c
      implicit none
      
c     ******************************************************************
c
c     input variables
c
c     ******************************************************************
c     from dims
      integer, intent(in) :: ny,il,jl,ie,je,ib,jb

c     from flo_var
      real(8), intent(inout), dimension(0:ib,0:jb,4) :: w
      real(8), intent(inout), dimension(0:ib,0:jb) :: p

c     from mesh_var
      real(8), intent(in), dimension(1:il,2:jl)   :: pori
      real(8), intent(in), dimension(2:il,1:jl)   :: porj

c     from solv_var
      real(8), intent(inout), dimension(0:ib,0:jb,4) :: fw
      real(8), intent(inout), dimension(0:ib,0:jb)   :: radi,radj

c     from solv_param
      real(8), intent(in)     :: rfil, vis2, vis4

c
c     ******************************************************************
c
c     local variables
c
c     ******************************************************************
c
      integer  :: i,j,n
c
c     ******************************************************************
c
      real(8)     :: fis2,fis4,sfil,plim,tol,rad
c
c     ******************************************************************
c
      real(8), dimension(0:ib,0:jb)   :: dp
      real(8), dimension(0:ie,0:je)   :: d
      real(8), dimension(il,jl)       :: e,gs
      real(8), dimension(il,jl)       :: dis2,dis4
c
c     ******************************************************************
c
      fis2      = 2.*rfil*vis2
      fis4      = rfil*vis4/16.
      sfil      = 1.  -rfil
      plim      = .001
      tol       = .25
c
c     replace the energy by the enthalpy
c
      do j=0,jb
      do i=0,ib
         w(i,j,4)  = w(i,j,4)  +p(i,j)
      end do
      end do
c
c     dissipation in the i direction
c
      do j=2,jl
         do i=1,ie
            dp(i,j)   = abs((p(i+1,j)  -2.*p(i,j)  +p(i-1,j))/
     .                      (p(i+1,j)  +2.*p(i,j)  +p(i-1,j)  +plim))
         end do
         dp(0,j)   = dp(1,j)
         dp(ib,j)  = dp(ie,j)
      end do

      do j=2,jl
      do i=1,il
         rad       = min(radi(i+1,j),radi(i,j))
         dis2(i,j) = fis2*rad*
     .               min(tol,max(dp(i+2,j),dp(i+1,j),dp(i,j),dp(i-1,j)))
         dis4(i,j) = fis4*rad
         dis4(i,j) = dim(dis4(i,j),dis2(i,j))
      end do
      end do

      do n=1,4

      do j=2,jl
      do i=0,ie
         d(i,j)    = w(i+1,j,n)  -w(i,j,n)
      end do
      end do

      do j=2,jl
      do i=1,il
         e(i,j)    = d(i+1,j)  -2.*d(i,j)  +d(i-1,j)
      end do
      end do

      do j=2,jl
      do i=1,il
         gs(i,j)   = pori(i,j)*(dis2(i,j)*d(i,j)  -dis4(i,j)*e(i,j))
      end do
      end do

      do j=2,jl
      do i=2,il
         fw(i,j,n) = sfil*fw(i,j,n)  +gs(i-1,j)  -gs(i,j)
      end do
      end do

      end do

      if (ny.lt.3) go to 11
c
c     dissipation in the j direction
c
      do j=1,je
      do i=2,il
         dp(i,j)   = abs((p(i,j+1)  -2.*p(i,j)  +p(i,j-1))/
     .                   (p(i,j+1)  +2.*p(i,j)  +p(i,j-1)  +plim))
      end do
      end do

      do i=2,il
         rad         = min(radj(i,2),radj(i,1))
         dis2(i,1)   = fis2*rad*
     .                 min(tol,max(dp(i,3),dp(i,2),dp(i,1)))
         dis4(i,1)   = fis4*rad
         dis4(i,1)   = dim(dis4(i,1),dis2(i,1))
         rad         = min(radj(i,jl+1),radj(i,jl))
         dis2(i,jl)  = fis2*rad*
     .                 min(tol,max(dp(i,jl+1),dp(i,jl),dp(i,jl-1)))
         dis4(i,jl)  = fis4*rad
         dis4(i,jl)  = dim(dis4(i,jl),dis2(i,jl))
      end do

      do j=2,ny
      do i=2,il
         rad       = min(radj(i,j+1),radj(i,j))
         dis2(i,j) = fis2*rad*
     .               min(tol,max(dp(i,j+2),dp(i,j+1),dp(i,j),dp(i,j-1)))
         dis4(i,j) = fis4*rad
         dis4(i,j) = dim(dis4(i,j),dis2(i,j))
      end do
      end do

      do n=1,4

      do j=0,je
      do i=2,il
         d(i,j)    = w(i,j+1,n)  -w(i,j,n)
      end do
      end do

      do j=1,jl
      do i=2,il
         e(i,j)    = d(i,j+1)  -2.*d(i,j)  +d(i,j-1)
      end do
      end do

      do j=1,jl
      do i=2,il
         gs(i,j)   = porj(i,j)*(dis2(i,j)*d(i,j)  -dis4(i,j)*e(i,j))
      end do
      end do

      do j=2,jl
      do i=2,il
         fw(i,j,n) = fw(i,j,n)  +gs(i,j-1)  -gs(i,j)
      end do
      end do

      end do
c
c     replace the enthalpy by the energy
c
   11 do j=0,jb
      do i=0,ib
         w(i,j,4)  = w(i,j,4)  -p(i,j)
      end do
      end do

      return

      end
